enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,358 @@
+
+## Motivation
+
+### Goals
+
+* Index multiple JSON formats defined by an application in Elasticsearch without indexing explosions.
+* Allow access to arbitrary JSON log message fields for non-Elasticsearch users.
+* Upgrade path from today's implementation that avoids extra downtime or risk of index explosions.
+
+### Non-Goals
+
+* General-purpose JSON queries and transformations.
+* Expression or validation of user-defined schema.
+
+## Proposal
+
+Structured logging (JSON parsing) is enabled in the ClusterLogForwarder CRD.
+It can be enabled at the `input` or `output` by adding a `structured` field.
+
+Enabling at the input applies a "format" name to selected logs.
+The format name can be used by the consumer to decide how to index or process the message.
+
+Enabling at the output constrains the forwarded record to be what the consumer expects.
+For example: to forward only structured messages (drop unstructured), or to flatten all messages into unstructured string form.
+
+Structured logging can be enabled at both input and output, or only at one.
+* If enabled at input, all connected outputs capable of receiving structured logs will do so.
+* If enabled at output, all connected inputs will be configured to do JSON parsing.
+
+See 'Proposal Details' below for configuration specifics.
+
+### User Stories
+
+Imagine there are two web server implementations named "patchy" and "bulldog".
+Each type of web server produces structured logs in a consistent format, but each uses a different format.
+A hypothetical admin deploys both patchy and bulldog servers.
+
+#### I want to index structured logs in Elasticsearch from applications in different namespaces with different formats
+
+Applications running in namespaces "fred" and "bob" use patchy.
+In namespaces "jill" and "jane" are applications using bulldog.
+
+The ClusterLogForwarder `input` selects logs and identifies their format.
+All logs are sent to the `default` Elasticsearch output.
+
+```yaml
+inputs:
+- name: patchyIn
+  application:
+    namespaces: [ fred, bob ]
+  structured:
+    format: patchy
+- name: bulldogIn
+  application:
+    namespaces: [ jill, jane ]
+  structured:
+    format: bulldog
+pipeline:
+- inputRefs: [ patchyIn, bulldogIn ]
+  outputRefs: default
+```
+
+Elasticsearch will create separate indices for "patchy and "bulldog" logs.
+See the [Elasticsearch notes](#es-notes) for more details on indexing.
+
+#### I want to index logs in multiple formats from one namespace or across multiple namespaces
+
+To do this, the admin must apply k8s labels to identify the pods using each format.
+Typically, this is _already done_ as part of a Deployment configuration.
+A Deployment normally labels pods by the applications that they contain.
+Those labels can also be used to identify the log format.
+
+**Note**: this requires implementation of [forwarder label selectors][clf-labels]
```

Comment:
Can we accelerate the work here so it is done when we release the JSON feature?

Reply:
Will try. In any case the JSON feature is designed so it can take advantage of present and future input selectors.